lib/backend/dir/impl.go

```diff
@@ -82,318 +82,306 @@ func New(params backend.Params) (backend.Backend, error) {
-func (bk *Backend) GetKeys(bucket []string) ([]string, error) {
-	files, err := ioutil.ReadDir(path.Join(bk.RootDir, path.Join(bucket...)))
+// GetItems returns all items (key/value pairs) in a given bucket.
+func (bk *Backend) GetItems(bucket []string) ([]backend.Item, error) {
+	var out []backend.Item
+
+	// Get a list of all buckets in the backend.
+	files, err := ioutil.ReadDir(path.Join(bk.rootDir))
 	if err != nil {
-		if os.IsNotExist(err) {
-			return []string{}, nil
-		}
 		return nil, trace.ConvertSystemError(err)
 	}
-	// enumerate all directory entries and select only non-hidden files
-	retval := make([]string, 0)
+
+	// Loop over all buckets in the backend.
 	for _, fi := range files {
-		name := fi.Name()
-		// legal keys cannot start with '.' (resrved prefix)
-		if name[0] != reservedPrefix {
-			retval = append(retval, name)
+		pathToBucket := bk.pathToBucket(fi.Name())
+		bucketPrefix := bk.flatten(bucket)
+
+		// Skip over any buckets without a matching prefix.
+		if !strings.HasPrefix(pathToBucket, bucketPrefix) {
+			continue
 		}
-	}
-	return retval, nil
-}
 
-// CreateVal creates value with a given TTL and key in the bucket
-// if the value already exists, returns AlreadyExistsError
-func (bk *Backend) CreateVal(bucket []string, key string, val []byte, ttl time.Duration) error {
-	// do not allow keys that start with a dot
-	if key[0] == reservedPrefix {
-		return trace.BadParameter("invalid key: '%s'. Key names cannot start with '.'", key)
-	}
-	// create the directory:
-	dirPath := path.Join(bk.RootDir, path.Join(bucket...))
-	err := os.MkdirAll(dirPath, defaultDirMode)
-	if err != nil {
-		return trace.ConvertSystemError(err)
-	}
-	// create the file (AKA "key"):
-	filename := path.Join(dirPath, key)
-	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_EXCL, defaultFileMode)
-	if err != nil {
-		if os.IsExist(err) {
-			return trace.AlreadyExists("%s/%s already exists", dirPath, key)
+		// Open the bucket to work on the items.
+		b, err := openBucket(pathToBucket, os.O_RDWR)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer b.Close()
+
+		// Loop over all keys, flatten them, and return key and value to caller.
+		for k, v := range b.items {
+			var key string
+
+			// If bucket path on disk and the requested bucket were an exact match,
+			// return the key as-is.
+			//
+			// However, if this was a partial match, for example pathToBucket is
+			// "/roles/admin/params" but the bucketPrefix is "/roles" then extract
+			// the first suffix (in this case "admin") and use this as the key. This
+			// is consistent with our DynamoDB implementation.
+			if pathToBucket == bucketPrefix {
+				key = k
+			} else {
+				key, err = suffix(pathToBucket, bucketPrefix)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+			}
+
+			// If the bucket item is expired, update the bucket, and don't include
+			// it in the output.
+			if bk.isExpired(v) {
+				delete(b.items, k)
```

Comment:
I would prefer this logic with`itemsUpdated` and `delete` operations to be encapsulated in the bucket, so 

`b.deleteItem(k)` should handle this internal property `itemsUpdated`